atlasdb-client/src/test/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueServiceTest.java

```diff
@@ -0,0 +1,230 @@
+        root.setLevel(loglevel);
+
+        final Appender mockAppender = mock(Appender.class);
+        when(mockAppender.getName()).thenReturn("MOCK");
+        root.addAppender(mockAppender);
+        return mockAppender;
+    }
+
+
+    @Before
+    public void before() throws Exception {
+        delegate = mock(KeyValueService.class);
+        kvs = ProfilingKeyValueService.create(delegate, 1000);
+    }
+
+    @After
+    public void after() throws Exception {
+        kvs.close();
+    }
+
+    @Test
+    public void dropTableSlowLogPresentOnInfoLevel() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.INFO);
+
+        doAnswer(waitASecond).when(delegate).dropTable(any());
+        kvs.dropTable(TABLE_REF);
+
+        verify(mockAppender).doAppend(slowLogMatcher.get());
+    }
+
+    @Test
+    public void dropTableTraceLogPresentOnTraceLevel() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.TRACE);
+
+        kvs.dropTable(TABLE_REF);
+
+        verify(mockAppender).doAppend(traceLogMatcher.get());
+    }
+
+    @Test
+    public void dropTableTraceLogPresentOnTraceLevelEvenIfQueryIsSlow() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.TRACE);
+
+        doAnswer(waitASecond).when(delegate).dropTable(any());
+        kvs.dropTable(TABLE_REF);
+
+        verify(mockAppender).doAppend(traceLogMatcher.get());
+        verify(mockAppender).doAppend(slowLogMatcher.get());
+    }
+
+    @Test
+    public void dropTableNoLoggingHappensIfQueryFastAndInfoLevel() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.INFO);
+
+        kvs.dropTable(TABLE_REF);
+
+        verifyNoMoreInteractions(mockAppender);
+    }
+
+    @Test
+    public void getAllTableNamesSlowLogPresentOnInfoLevel() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.INFO);
+
+        doAnswer(waitASecond).when(delegate).getAllTableNames();
+        kvs.getAllTableNames();
+
+        verify(mockAppender).doAppend(slowLogMatcher.get());
+    }
+
+    @Test
+    public void getAllTableNamesTraceLogPresentOnTraceLevel() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.TRACE);
+
+        kvs.getAllTableNames();
+
+        verify(mockAppender).doAppend(traceLogMatcher.get());
+    }
+
+    @Test
+    public void getAllTableNamesTraceLogPresentOnTraceLevelEvenIfQueryIsSlow() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.TRACE);
+
+        doAnswer(waitASecond).when(delegate).getAllTableNames();
+        kvs.getAllTableNames();
+
+        verify(mockAppender).doAppend(traceLogMatcher.get());
+        verify(mockAppender).doAppend(slowLogMatcher.get());
+    }
+
+    @Test
+    public void getAllTableNamesNoLoggingHappensIfQueryFastAndInfoLevel() {
+        Appender mockAppender = setLogLevelAndGetAppender(Level.INFO);
+
+        kvs.getAllTableNames();
+
+        verifyNoMoreInteractions(mockAppender);
```

Comment:
There is a small but unlikely chance of a flake here depending on the load on the CI at the time of testing.  Let's keep an eye on it for now.